File: game/vector_field.py
"""Creates a vector field useful for navigating enemies around the game map."""
from __future__ import annotations

# Builtin
import logging
import time
from collections import deque
from typing import TYPE_CHECKING

# Pip
import numpy as np

# Custom
from game.common import grid_bfs
from game.constants.game_object import SPRITE_SIZE

if TYPE_CHECKING:
    import arcade

__all__ = ("VectorField",)

# Get the logger
logger = logging.getLogger(__name__)


intercardinal_offsets: tuple[tuple[int, int], ...] = (
    (-1, -1),
    (0, -1),
    (1, -1),
    (-1, 0),
    (1, 0),
    (-1, 1),
    (0, 1),
    (1, 1),
)


class VectorField:
    """Represents a vector field allowing pathfinding for large amounts of enemies.

    The steps needed to accomplish this:
        1. First, we start at the destination tile and work our way outwards using a
        breadth first search. This is called a 'flood fill' and will construct the
        Dijkstra map needed for the vector field.

        2. Next, we iterate over each tile and find the neighbour with the lowest
        Dijkstra distance. Using this we can create a vector from the source tile to the
        neighbour tile making for more natural pathfinding since the enemy can go in 6
        directions instead of 4.

        3. Finally, once the neighbour with the lowest Dijkstra distance is found, we
        can create a vector from the current tile to that neighbour tile which the enemy
        will follow. Repeating this for every tile gives us an efficient way to
        calculate pathfinding for a large amount of entities.

    Further reading which may be useful:
    `Other uses of Dijkstra maps
    <http://www.roguebasin.com/index.php/The_Incredible_Power_of_Dijkstra_Maps>`_
    `Dijkstra maps visualized
    <http://www.roguebasin.com/index.php/Dijkstra_Maps_Visualized>`_
    `Understanding goal based pathfinding
    <https://gamedevelopment.tutsplus.com/tutorials/understanding-goal-based-vector\
    -field-pathfinding--gamedev-9007>`_

    Parameters
    ----------
    width: int
        The width of the grid.
    height: int
        The height of the grid.
    walls: arcade.SpriteList
        A list of wall sprites that can block the entities.

    Attributes
    ----------
    walls_dict: dict[tuple[int, int], int]
        A dictionary which persistently holds the wall tile positions and their
        distances from the destination tile. This is used to update the main distances
        dict when the vector field is recalculated.
    distances: dict[tuple[int, int], int]
        A dictionary which holds a tuple containing the tile positions and its distance
        from the destination tile.
    vector_dict: dict[tuple[int, int], tuple[float, float]]
        A dictionary which holds a tuple containing the tile position and the vector the
        enemy should move on that when on that tile.
    """

    __slots__ = (
        "width",
        "height",
        "walls_dict",
        "distances",
        "vector_dict",
    )

    def __init__(
        self,
        width: int,
        height: int,
        walls: arcade.SpriteList,
    ) -> None:
        self.width: int = width
        self.height: int = height
        self.walls_dict: dict[tuple[int, int], int] = {}
        self.distances: dict[tuple[int, int], int] = {}
        self.vector_dict: dict[tuple[int, int], tuple[float, float]] = {}
        for wall in walls:
            self.walls_dict[self.get_tile_pos_for_pixel(wall.position)] = np.inf

    def __repr__(self) -> str:
        """Return a human-readable representation of this object."""
        return f"<VectorField (Width={self.width}) (Height={self.height})>"

    @staticmethod
    def get_tile_pos_for_pixel(position: tuple[float, float]) -> tuple[int, int]:
        """Calculate the tile position from a given screen position.

        Parameters
        ----------
        position: tuple[float, float]
            The sprite position on the screen.

        Raises
        ------
        ValueError
            The inputs must be bigger than or equal to 0.

        Returns
        -------
        tuple[int, int]
            The path field grid tile position for the given sprite position.
        """
        # Check if the inputs are negative
        x, y = position
        if x < 0 or y < 0:
            raise ValueError("The inputs must be bigger than or equal to 0.")

        # Calculate the grid position
        return int(x // SPRITE_SIZE), int(y // SPRITE_SIZE)

    def recalculate_map(self, player_pos: tuple[float, float]) -> None:
        """Recalculates the vector field and produces a new path_dict.

        Parameters
        ----------
        player_pos: tuple[float, float]
            The position of the player on the screen.
        """
        # Record the start time, so we can know how long the generation takes
        start_time = time.perf_counter()

        # To recalculate the map, we need a few things:
        #   1. A distances dict to store the distances to each tile position from the
        #   destination tile position. This needs to only include the elements inside
        #   the walls dict.
        #   2. A vector_dict dict to store the paths for the vector field. We also need
        #   to make sure this is empty first.
        #   3. A queue object, so we can explore the grid.
        start = self.get_tile_pos_for_pixel(player_pos)
        self.distances.clear()
        self.distances |= self.walls_dict
        self.distances[start] = 0
        self.vector_dict.clear()
        queue: deque[tuple[int, int]] = deque[tuple[int, int]]()
        queue.append(start)

        # Explore the grid using a breadth first search to generate the Dijkstra
        # distances
        while bool(queue):
            # Get the current tile to explore
            current = queue.popleft()

            # Sometimes current can be None, so check if it is None
            if not current:  # pragma: no cover
                continue

            # Get the current tile's neighbours
            for neighbour in grid_bfs(current, self.height, self.width):
                # Check if the neighbour is a wall or not
                if self.distances.get((neighbour[0], neighbour[1]), -1) == np.inf:
                    continue

                # Test if the neighbour has already been reached or not. If it hasn't,
                # add it to the queue and set its distance
                if neighbour not in self.distances:
                    queue.append(neighbour)
                    self.distances[neighbour] = 1 + self.distances[current]

        # Use the newly generated Dijkstra map to get a neighbour with the lowest
        # Dijkstra distance at each tile. Then create a vector pointing in that
        # direction
        for tile, cost in self.distances.items():
            # If this tile is a wall tile, ignore it
            if cost == np.inf:
                continue

            # Find the tile's neighbour with the lowest Dijkstra distance
            min_tile = -1, -1
            min_dist = np.inf
            for neighbour in grid_bfs(
                tile, self.width, self.height, offsets=intercardinal_offsets
            ):
                # Sometimes an invalid tile is returned so test for that
                distance = self.distances.get(neighbour, -1)
                if distance < min_dist and distance != -1:
                    min_tile = neighbour
                    min_dist = distance

            # Now point the tile's vector in the direction of the tile with the lowest
            # Dijkstra distance
            self.vector_dict[tile] = -(tile[0] - min_tile[0]), -(tile[1] - min_tile[1])

        # Set the vector for the destination tile to avoid weird movement when the enemy
        # is touching the player
        self.vector_dict[start] = 0, 0

        # Log the time taken to generate the vector field
        logger.debug(
            "Vector field generated in %f seconds", time.perf_counter() - start_time
        )

    def get_vector_direction(
        self, current_enemy_pos: tuple[float, float]
    ) -> tuple[float, float]:
        """Get the vector the enemy should travel on based on their current position.

        Parameters
        ----------
        current_enemy_pos: tuple[float, float]
            The current position of the enemy on the screen.

        Returns
        -------
        tuple[float, float[
            The vector the enemy needs to travel in.
        """
        return self.vector_dict[self.get_tile_pos_for_pixel(current_enemy_pos)]

File: hades/extensions/src/vector_field.cpp
#include "hades_common.h"
#include <structmember.h>
#include <deque>
#include <string>
#include <stdexcept>
#include <iostream>


// ----- CONSTANTS ------------------------------
/* The width, height and sprite size constants for use in the vector field */
int WIDTH, HEIGHT;
int SPRITE_SIZE = 56;


/* An unordered map of wall grid positions with the cost of traversing them. This is
used to update the distances unordered_map in recalculate_map() since the cost of
traversing a wall is always infinity */
std::unordered_map<IntPair, int> WALLS_DICT;


// ----- DOCSTRINGS ------------------------------
/* Stores the docstring for the vector_field module */
PyDoc_STRVAR(
    vector_field_module_docstring,
    "Creates a vector field useful for navigating enemies around the game map."
);


/* Stores the docstring for the VectorField type */
PyDoc_STRVAR(
    vector_field_type_docstring,
    "Represents a vector field allowing pathfinding for large amounts of enemies.\n\n"
    "The steps needed to accomplish this:\n"
    "    1. First, we start at the destination tile and work our way outwards using a"
    "    breadth first search. This is called a 'flood fill' and will construct the"
    "    Dijkstra map needed for the vector field.\n"
    "    2. Next, we iterate over each tile and find the neighbour with the lowest"
    "    Dijkstra distance. Using this we can create a vector from the source tile to"
    "    the neighbour tile making for more natural pathfinding since the enemy can go"
    "    in 6 directions instead of 4.\n"
    "    3. Finally, once the neighbour with the lowest Dijkstra distance is found, we"
    "    can create a vector from the current tile to that neighbour tile which the"
    "    enemy will follow. Repeating this for every tile gives us an efficient way to"
    "    calculate pathfinding for a large amount of entities.\n\n"
    "Further reading which may be useful:\n"
    "`Other uses of Dijkstra maps"
    "<http://www.roguebasin.com/index.php/The_Incredible_Power_of_Dijkstra_Maps>`_\n"
    "`Dijkstra maps visualized"
    "<http://www.roguebasin.com/index.php/Dijkstra_Maps_Visualized>`_\n"
    "`Understanding goal based pathfinding"
    "<https://gamedevelopment.tutsplus.com/tutorials/understanding-goal-based-vector/-"
    "field-pathfinding--gamedev-9007>`_\n\n"
    "Parameters\n"
    "----------\n"
    "walls: arcade.SpriteList\n"
    "    A list of wall sprites that can block the entities.\n"
    "width: int\n"
    "    The width of the grid.\n"
    "height: int\n"
    "    The height of the grid.\n\n"
    "Attributes\n"
    "----------\n"
    "vector_dict: dict[tuple[int, int], tuple[float, float]]\n"
    "    A dictionary which holds a tuple containing the tile position and the vector"
    "    the enemy should follow when on that tile."
);

/* Stores the docstring for the VectorField.pixel_to_tile_pos method */
PyDoc_STRVAR(
    pixel_to_tile_pos_docstring,
    "Calculate the tile position from a given screen position.\n\n"
    "Parameters\n"
    "----------\n"
    "position: tuple[float, float]\n"
    "    The sprite position on the screen.\n\n"
    "Raises\n"
    "------\n"
    "    The inputs must be bigger than or equal to 0.\n\n"
    "Returns\n"
    "-------\n"
    "    The path field grid tile position for the given sprite position."
);

/* Stores the docstring for the VectorField.recalculate_map method */
PyDoc_STRVAR(
    recalculate_map_docstring,
    "Recalculates the vector field and produces a new path_dict.\n\n"
    "Parameters\n"
    "----------\n"
    "player_pos: tuple[float, float]\n"
    "    The position of the player on the screen.\n"
    "player_view_distance: int\n"
    "    The player's view distance.\n\n"
    "Returns\n"
    "-------\n"
    "list[tuple[int, int]]\n"
    "    A list of the possible enemy spawns."
);


// ----- C STRUCTURE DEFINITIONS ------------------------------
struct FloatPair {
    /* Represents a namedtuple describing a pair of floats */
    float x, y;
};


typedef struct {
    /* Represents a VectorField type and what it contains */
    PyObject base;
    PyObject *vector_dict;
} VectorField;


// ----- C METHODS ------------------------------
IntPair get_tile_pos_for_pixel(FloatPair screen_position) {
    /* Calculates the tile position from a given screen position */
    return {(int) floor(screen_position.x / SPRITE_SIZE), (int) floor(screen_position.y / SPRITE_SIZE)};
}


// ----- VECTORFIELD METHODS ------------------------------
static PyObject *pixel_to_tile_pos(PyObject *self, PyObject *args) {
    /* Parse arguments */
    FloatPair position;
    if (!PyArg_ParseTuple(args, "(ff)", &position.x, &position.y)) {
        return Py_BuildValue("");
    }

    /* Converting logic */
    IntPair tile_pos = get_tile_pos_for_pixel(position);

    /* Return result */
    return Py_BuildValue("(ii)", tile_pos.x, tile_pos.y);
}


static PyObject *recalculate_map(PyObject *obj, PyObject *args) {
    /* Parse arguments */
    VectorField *self = (VectorField*) obj;
    FloatPair player_pos;
    int player_view_distance;
    if (!PyArg_ParseTuple(args, "(ff)i", &player_pos.x, &player_pos.y, &player_view_distance)) {
        return Py_BuildValue("");
    }

    /* Vector field recalculation logic */
    // To recalculate the map, we need a few things:
    //      1. A distances dict to store the distances to each tile position from the
    //      destination tile position. This needs to only include the elements inside
    //      the walls dict.
    //      2. A vector_dict dict to store the paths for the vector field. We also need
    //      to make sure this is empty first.
    //      3. A queue object, so we can explore the grid.
    //      4. A possible_spawns list to hold all the grid positions where enemies can
    //      spawn.
    IntPair start = get_tile_pos_for_pixel(player_pos);
    std::unordered_map<IntPair, int> distances = WALLS_DICT;
    distances[start] = 0;
    PyDict_Clear(self->vector_dict);
    PyObject *possible_spawns = PyList_New(0);
    std::deque<IntPair> queue = {start};

    // Explore the grid using a breadth first search to generate the Dijkstra distances
    while (!queue.empty()) {
        // Get the current tile to explore
        IntPair current = queue.front();
        queue.pop_front();

        // Get the current tile's neighbours
        for (IntPair neighbour: grid_bfs(current, HEIGHT, WIDTH)) {
            // Check if the neighbour is a wall or not. If so, continue to the next
            // iteration. However, if it doesn't exist, add it to the queue and set its
            // distance
            try {
                if (distances.at(neighbour) == INT_INFINITY) {
                    continue;
                }
            } catch (std::out_of_range) {
                queue.push_back(neighbour);
                distances[neighbour] = 1 + distances[current];
            }
        }
    }

    // Use the newly generated Dijkstra map to get a neighbour with the lowest Dijkstra
    // distance at each tile. Then create a vector pointing in that direction
    for (std::pair<IntPair, int> tile : distances) {
        // If this tile is a wall tile, ignore it
        if (tile.second == INT_INFINITY) {
            continue;
        }

        // Find the tile's neighbour with the lowest Dijkstra distance
        PyObject *py_tile = Py_BuildValue("(ii)", tile.first.x, tile.first.y);
        IntPair min_tile = {0, 0};
        int min_dist = INT_INFINITY;
        for (IntPair neighbour : grid_bfs(tile.first, HEIGHT, WIDTH, INTERCARDINAL_OFFSETS)) {
            // Test if the neighbour has the lowest Dijkstra distance
            if (distances[neighbour] < min_dist) {
                min_tile = neighbour;
                min_dist = distances[neighbour];
            }
        }

        // Now point the tile's vector in the direction of the tile with the lowest
        // Dijkstra distance
        PyDict_SetItem(self->vector_dict, py_tile, Py_BuildValue("(ii)", -(tile.first.x - min_tile.x), -(tile.first.y - min_tile.y)));

        // Also test if the current tile is outside the player's fov
        if (tile.second > player_view_distance) {
            PyList_Append(possible_spawns, py_tile);
        }
    }

    // Set the vector for the destination tile to avoid weird movement when the enemy
    // is touching the player
    PyDict_SetItem(self->vector_dict, Py_BuildValue("(ii)", start.x, start.y), Py_BuildValue("(ii)", 0, 0));

    /* Return result */
    return possible_spawns;
}


// ----- VECTORFIELD MAGIC METHODS ------------------------------
static PyObject *vector_field_repr(VectorField *self) {
    /* Build repr */
    std::string repr = "<VectorField (Width=" + std::to_string(WIDTH) + ") (Height=" + std::to_string(HEIGHT) + ")>";

    /* Return result */
    return Py_BuildValue("s", repr.c_str());
}


static int vector_field_init(VectorField *self, PyObject *args, PyObject *kwds) {
    /* Parse arguments */
    PyObject *walls;
    if (!PyArg_ParseTuple(args, "Oii", &walls, &WIDTH, &HEIGHT)) {
        return -1;
    }

    /* VectorField initialisation logic */
    // Convert each sprite's position in the spritelist to a grid position and store
    // it in walls_dict with infinity as the value
    self->vector_dict = PyDict_New();
    PyObject *sprite_list = PyObject_GetAttrString(walls, "sprite_list");
    Py_DECREF(walls);
    for (int i = 0; i < PyList_Size(sprite_list); i++) {
        PyObject *py_screen_position = PyObject_GetAttrString(PyList_GetItem(sprite_list, i), "position");
        float x = (float) PyFloat_AsDouble(PyTuple_GetItem(py_screen_position, 0));
        float y = (float) PyFloat_AsDouble(PyTuple_GetItem(py_screen_position, 1));
        WALLS_DICT[get_tile_pos_for_pixel({x, y})] = INT_INFINITY;
    }
    Py_DECREF(sprite_list);

    /* Return result */
    return 0;
}


// ----- VECTORFIELD DEFINITIONS ------------------------------
static PyMethodDef vector_field_methods[] = {
    /* Defines the methods which belong to the VectorField type */
    {"pixel_to_tile_pos", pixel_to_tile_pos, METH_VARARGS | METH_STATIC, pixel_to_tile_pos_docstring},
    {"recalculate_map", recalculate_map, METH_VARARGS},
    {NULL},
};


static PyMemberDef vector_field_members[] = {
    /* Defines the members which belong to the VectorField type */
    {"vector_dict", T_OBJECT_EX, offsetof(VectorField, vector_dict), READONLY},
    {NULL},
};


static PyTypeObject VectorFieldType = {
    /* Describes the VectorField type and all of its slots */
    PyVarObject_HEAD_INIT(NULL, 0)
    "vector_field.VectorField",     // tp_name
    sizeof(VectorField),            // tp_basicsize
    0,                              // tp_itemsize
    0,                              // tp_dealloc
    0,                              // tp_vectorcall_offset
    0,                              // tp_getattr
    0,                              // tp_setattr
    0,                              // tp_as_async
    (reprfunc) vector_field_repr,   // tp_repr
    0,                              // tp_as_number
    0,                              // tp_as_sequence
    0,                              // tp_as_mapping
    0,                              // tp_hash
    0,                              // tp_call
    0,                              // tp_str
    0,                              // tp_getattro
    0,                              // tp_setattro
    0,                              // tp_as_buffer
    Py_TPFLAGS_DEFAULT,             // tp_flags
    vector_field_type_docstring,    // tp_doc
    0,                              // tp_traverse
    0,                              // tp_clear
    0,                              // tp_richcompare
    0,                              // tp_weaklistoffset
    0,                              // tp_iter
    0,                              // tp_iternext
    vector_field_methods,           // tp_methods
    vector_field_members,           // tp_members
    0,                              // tp_getset
    0,                              // tp_base
    0,                              // tp_dict
    0,                              // tp_descr_get
    0,                              // tp_descr_set
    0,                              // tp_dictoffset
    (initproc) vector_field_init,   // tp_init
    0,                              // tp_alloc
    PyType_GenericNew,              // tp_new
};


// ----- MODULE DEFINITION ------------------------------
static struct PyModuleDef vector_field_module = {
    /* Defines the metadata for this extension module */
    PyModuleDef_HEAD_INIT,
    "vector_field",
    vector_field_module_docstring,
    -1,
    NULL,
};


// ----- MODULE CREATION ------------------------------
PyMODINIT_FUNC PyInit_vector_field(void) {
    /* Initialises this module so Python can access it */
    // Initialise the C++ module and check if it's valid or not
    PyObject *module = PyModule_Create(&vector_field_module);
    if (module == NULL)
        return NULL;

    // Initialise the constants
    SPRITE_SIZE = (int) PyFloat_AsDouble(get_global_constant("hades.constants.game_object", {"SPRITE_SIZE"}));

    // Initialise the VectorField object
    if (PyModule_AddType(module, &VectorFieldType) < 0) {
        return NULL;
    }

    // Return the C++ initialised module
    return module;
}
